[JSC] DFG GetByOffset / PutByOffset should be optimized with Double

Constellation opened the pull request:
#### 7ddf52c48855ebaaa302fae8fe5bce0ce223897b
<pre>
[JSC] DFG GetByOffset / PutByOffset should be optimized with Double
<a href="https://bugs.webkit.org/show_bug.cgi?id=296666">https://bugs.webkit.org/show_bug.cgi?id=296666</a>
<a href="https://rdar.apple.com/157062331">rdar://157062331</a>

Reviewed by Yijia Huang.

This patch implements speculative double load / store via GetByOffset /
PutByOffset in DFG too. It is already implemented in FTL, but this
extends that optimization for DFG.

While FTL identifies these opportunities based on ValueRep reduction
phase, DFG does it more speculatively in DFG fixup phase. The reason is
that FTL needs to do object allocation sinking and argument elimination
after the fixup, and making them DoubleResult introduces complicated
problem for these phases. So FTL handles it in a different way:
identifying them in ValueRep reduction phase, which is more
sophisticated.
But DFG does not have these phases (including ValueRep reduction phase),
thus we do simply speculatively in fixup phase based on prediction
information.

The core of the optimization is the same to FTL&apos;s one. We load value
directly to the FPR, and extract double by doing integer arithmetics
onto the FPR. We avoid using fmov between GPRs and FPRs, which is really
costly.

This patch also does the same thing for ClosureVar and GlobalVars.

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::storeDouble):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::unboxRealNumberDouble):
(JSC::DFG::SpeculativeJIT::boxDoubleAsDouble):
* Source/JavaScriptCore/dfg/DFGVarargsForwardingPhase.cpp:

Canonical link: <a href="https://commits.webkit.org/298092@main">https://commits.webkit.org/298092@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f73f31eae918e230c631e77993388dd4c4f2184

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114225 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33972 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24434 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120391 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/64961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/da134e2a-b0e6-4053-867b-f97bb6175623) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34601 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42534 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/86804 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/64961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/06c4c9d6-0462-4bc8-b6c5-4c26e9443ab2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117173 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27552 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102588 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67193 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26734 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20714 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64086 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/106630 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/96926 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/20831 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123608 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/112788 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41244 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30752 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/95643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41621 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98790 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/95426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40560 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18387 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18304 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41123 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46634 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/136985 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40730 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36666 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44036 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42481 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
```diff
@@ -2112,15 +2112,64 @@ class FixupPhase : public Phase {
+                }
+            }
+#endif
+            break;
```
nit: maybe refactor to a `shouldSpeculateDouble` function?
```
bool shouldSpeculateDouble(Node* node, Edge& edge)
{
#if USE(JSVALUE64)
    return !m_graph.m_plan.isFTL()
        && !m_graph.hasExitSite(node->origin.semantic, BadType)
        && !edge->shouldSpeculateInt32()
        && edge->shouldSpeculateNumber();
#else
    UNUSED_PARAM(node);
    UNUSED_PARAM(edge);
    return false;
#endif
}
```
uses
```
if (shouldSpeculateDouble(node, node->child3())) {
    fixEdge<DoubleRepUse>(node->child3());
    break;
}
```
and
```
if (shouldSpeculateDouble(node, node)) {
    node->setResult(NodeResultDouble);
    break;
}
```

Constellation replied:
That sounds nice! Changed.

Constellation replied:
I think
```
    return !m_graph.m_plan.isFTL()
        && !m_graph.hasExitSite(node->origin.semantic, BadType)
```
these things should be kept in the caller side for readability (why we don't want to do that, and how these exit sites are attached to what node type).

Constellation replied:
I may add `shouldSpeculateDoubleResult`.

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] DFG GetByOffset / PutByOffset should be optimized with Double
https://bugs.webkit.org/show_bug.cgi?id=296666
rdar://157062331

Reviewed by Yijia Huang.

This patch implements speculative double load / store via GetByOffset /
PutByOffset in DFG too. It is already implemented in FTL, but this
extends that optimization for DFG.

While FTL identifies these opportunities based on ValueRep reduction
phase, DFG does it more speculatively in DFG fixup phase. The reason is
that FTL needs to do object allocation sinking and argument elimination
after the fixup, and making them DoubleResult introduces complicated
problem for these phases. So FTL handles it in a different way:
identifying them in ValueRep reduction phase, which is more
sophisticated.
But DFG does not have these phases (including ValueRep reduction phase),
thus we do simply speculatively in fixup phase based on prediction
information.

The core of the optimization is the same to FTL's one. We load value
directly to the FPR, and extract double by doing integer arithmetics
onto the FPR. We avoid using fmov between GPRs and FPRs, which is really
costly.

This patch also does the same thing for ClosureVar and GlobalVars.

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::storeDouble):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::unboxRealNumberDouble):
(JSC::DFG::SpeculativeJIT::boxDoubleAsDouble):
* Source/JavaScriptCore/dfg/DFGVarargsForwardingPhase.cpp:

Canonical link: https://commits.webkit.org/298092@main
